onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2686,6 +2686,149 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
+          "Constrains output type to all optional tensor or optional sequence types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        const auto* attr_proto = ctx.getAttribute("type");
+        if (attr_proto == nullptr)
+          fail_type_inference(
+                "Attribute type should be provided.");
+        if (!attr_proto->has_tp())
+          fail_type_inference(
+              "Attribute type should be of type TypeProto and specify a type.");
+        auto attr_tp = attr_proto->tp();
+        const size_t numOutputs = ctx.getNumOutputs();
+        if (numOutputs != 1) {
+          fail_type_inference("OptionalEmpty is expected to have 1 output.");
+        }
+        ctx.getOutputType(0)
+            ->mutable_optional_type()
+            ->mutable_elem_type()
+            ->CopyFrom(attr_tp);
+      });
+
+  static const char* OptionalConstruct_ver1_doc = R"DOC(
+      Construct a optional type containing the 'input' element.
+      )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OptionalConstruct)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(OptionalConstruct_ver1_doc)
+      .Input(0, "input", "The input element.", "T")
+      .Output(0, "output", "The optional output enclosing the input element.", "O")
+      .TypeConstraint(
+          "T",
+          {"tensor(float)",
+           "seq(tensor(float))"},
+          "Constrains input type to all tensor and sequence types.")
+      .TypeConstraint(
+          "O",
+          {"optional(tensor(float))",
+           "optional(seq(tensor(float)))"},
+          "Constrains output type to all optional tensor or optional sequence types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          const size_t numInputs = ctx.getNumInputs();
+          if (numInputs != 1) {
+            fail_type_inference("OptionalConstruct is expected to have 1 input.");
+          }
+          auto input_type = ctx.getInputType(0);
+          if(input_type == nullptr){
+            fail_type_inference("Input type is null. Type info is expected.");
+          }
+          const size_t numOutputs = ctx.getNumOutputs();
+          if (numOutputs != 1) {
+            fail_type_inference("OptionalConstruct is expected to have 1 output.");
+          }
+          ctx.getOutputType(0)
+              ->mutable_optional_type()
+              ->mutable_elem_type()
+              ->CopyFrom(*input_type);
+        });
+
+  static const char* OptionalHasElement_ver1_doc = R"DOC(
+      Construct a optional type containing the 'input' element.
+      )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OptionalHasElement)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(OptionalHasElement_ver1_doc)
+      .Input(0, "input", "The optional input.", "O")
+      .Output(0, "output", "A boolean tensor of empty shape. If true, it indicates that optional input contains a tensor or sequence element. Otherwise, it is empty.", "B")
```

Comment:
Nit: change "tensor of empty shape" to "scalar (that is, a tensor of rank zero)".

Reply:
nit: May be better to change "contains a tensor or sequence element" to "contains an element". It will continue to work even if we extend the type-constraint to allow other types in the future.